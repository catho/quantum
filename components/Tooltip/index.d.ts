import { FC, ReactNode } from 'react';

export interface TooltipProps {
    text: ReactNode;
    placement?: 'top' | 'right' | 'bottom' | 'left';
    visible?: boolean;
    multiline?: boolean;
    children: ReactNode[] | ReactNode;
    theme?: {
        spacing?: {};
        colors?: {};
        baseFontSize?: number;
    };
}

declare const Tooltip: FC<TooltipProps>;
export default Tooltip;
