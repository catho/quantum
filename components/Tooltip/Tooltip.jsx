import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import placementConfig from './options';
import {
  colors,
  spacing,
  baseFontSize as defaultBaseFontSize,
} from '../shared/theme';

const Tip = styled.div`
  border-radius: 4px;
  font-weight: bold;
  opacity: ${({ visible }) => (visible ? '1' : '0')};
  position: absolute;
  line-height: 0;
  text-align: center;
  transition: opacity 0.2s ease-in-out, visibility 0.2s ease-in-out;
  z-index: 100;

  ${({
    theme: {
      colors: { neutral },
      spacing: { xsmall },
      baseFontSize,
    },
  }) => `
    background-color: ${neutral[700]};
    border-color: ${neutral[700]};
    color: ${neutral[100]};
    font-size: ${baseFontSize}px;
    padding: ${xsmall}px;
  `}

  ${({ placement }) => placementConfig.tipPosition[placement]};

  &:before {
    content: '';
    position: absolute;
    ${({ placement }) => placementConfig.arrowPosition[placement]};
  }
`;

const TipText = styled.span`
  display: inline-block;
  max-width: 250px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  line-height: 20px;
`;

const Wrapper = styled.div`
  position: relative;
  float: left;
  clear: left;
`;

class Tooltip extends Component {
  constructor(props) {
    super(props);
    const { visible } = this.props;
    this.state = { visible };
  }

  isVisible = visible => this.setState({ visible });

  render() {
    const {
      children,
      placement,
      text,
      visible: visibleProp,
      theme,
      ...rest
    } = this.props;
    const { visible: visibleState } = this.state;

    return (
      <Wrapper
        onMouseEnter={() => this.isVisible(true)}
        onMouseLeave={() => this.isVisible(false)}
        {...rest}
      >
        <Tip
          placement={placement}
          visible={visibleProp || visibleState}
          theme={theme}
        >
          <TipText>{text}</TipText>
        </Tip>
        {children}
      </Wrapper>
    );
  }
}

Tip.displayName = 'Tip';

Tooltip.propTypes = {
  /** Text that tooltip will show */
  text: PropTypes.string.isRequired,
  /** Define tooltip positioning */
  placement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  visible: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  theme: PropTypes.shape({
    spacing: PropTypes.object,
    colors: PropTypes.object,
    baseFontSize: PropTypes.number,
  }),
};

Tooltip.defaultProps = {
  placement: 'top',
  visible: false,
  theme: {
    spacing,
    colors,
    baseFontSize: defaultBaseFontSize,
  },
};

export default Tooltip;
