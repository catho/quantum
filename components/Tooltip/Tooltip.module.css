:root {
    --tip-arrow-size: 6px;
    --tip-box-margin: 6px;
    --tip-percentage-x: 50%;
    --tip-percentage-y: 100%;
    --tip-position-distance: calc(var(--tip-percentage-y) + var(--tip-arrow-size) + var(--tip-box-margin));
}

.tooltip-wrapper {
    position: relative;
    float: left;
    clear: left;
    width: unset;
}

.tooltip-wrapper-multiline {
    width: max-content;
}

.tip {
    border-radius: 4px;
    font-weight: bold;
    opacity: 0;
    visibility: hidden;
    position: absolute;
    line-height: 0;
    text-align: center;
    transition:
        opacity 0.2s ease-in-out,
        visibility 0.2s ease-in-out;
    z-index: 100;
    background-color: var(--qtm-colors-neutral-700);
    border-color: var(--qtm-colors-neutral-700);
    color: var(--qtm-colors-neutral-100);
    font-size: var(--qtm-base-font-size);
    padding: var(--qtm-spacing-xsmall);
}

.tip-top {
    left: var(--tip-percentage-x);
    bottom: var(--tip-position-distance);
    transform: translateX(calc(var(--tip-percentage-x) * -1));
}

.tip-bottom {
    left: var(--tip-percentage-x);
    top: var(--tip-position-distance);
    transform: translateX(calc(var(--tip-percentage-x) * -1));
}

.tip-left {
    right: var(--tip-position-distance);
    top: var(--tip-percentage-x);
    transform: translateY(calc(var(--tip-percentage-x) * -1));
}

.tip-right {
    left: var(--tip-position-distance);
    top: var(--tip-percentage-x);
    transform: translateY(calc(var(--tip-percentage-x) * -1));
}

.tip:before {
    content: '';
    position: absolute;
}

.tip-top:before {
    border-left: var(--tip-arrow-size) solid transparent;
    border-right: var(--tip-arrow-size) solid transparent;
    left: 50%;
    transform: translateX(-50%);

    border-top: var(--tip-arrow-size) solid;
    border-top-color: inherit;
    bottom: calc((var(--tip-arrow-size) - 1px) * -1);
}

.tip-bottom:before {
    border-left: var(--tip-arrow-size) solid transparent;
    border-right: var(--tip-arrow-size) solid transparent;
    left: 50%;
    transform: translateX(-50%);

    border-bottom: var(--tip-arrow-size) solid;
    border-bottom-color: inherit;
    top: calc((var(--tip-arrow-size) - 1px) * -1);
}

.tip-right:before {
    border-top: var(--tip-arrow-size) solid transparent;
    border-bottom: var(--tip-arrow-size) solid transparent;
    top: 50%;
    transform: translateY(-50%);

    border-right: var(--tip-arrow-size) solid;
    border-right-color: inherit;
    left: calc((var(--tip-arrow-size) - 1px) * -1);
}

.tip-left:before {
    border-top: var(--tip-arrow-size) solid transparent;
    border-bottom: var(--tip-arrow-size) solid transparent;
    top: 50%;
    transform: translateY(-50%);

    border-left: var(--tip-arrow-size) solid;
    border-left-color: inherit;
    right: calc((var(--tip-arrow-size) - 1px) * -1);
}

.tip-visible {
    opacity: 1;
    visibility: visible;
}

.tip-text {
    display: inline-block;
    max-width: 250px;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    line-height: 20px;
    text-align: unset;
}

.tip-text-multiline {
    max-width: unset;
    white-space: pre;
    text-align: left;
}