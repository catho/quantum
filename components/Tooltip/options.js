const ARROW_SIZE = 6;
const PERCENTAGE_X = 50;
const BOX_MARGIN = 6;
const PERCENTAGE_Y = 100;
const tipPositionDistance = `calc(${PERCENTAGE_Y}% + ${ARROW_SIZE +
  BOX_MARGIN}px);`;

const upDownBorders = `
  border-left: ${ARROW_SIZE}px solid transparent;
  border-right: ${ARROW_SIZE}px solid transparent;
  left: 50%;
  transform: translateX(-50%);
`;

const sideBorders = `
  border-top: ${ARROW_SIZE}px solid transparent;
  border-bottom: ${ARROW_SIZE}px solid transparent;
  top: 50%;
  transform: translateY(-50%);
`;

const placementConfig = {
  arrowPosition: {
    top: `
      ${upDownBorders}
      border-top: ${ARROW_SIZE}px solid;
      border-top-color: inherit;
      bottom: -${ARROW_SIZE - 1}px;
    `,
    right: `
    ${sideBorders}
    border-right: ${ARROW_SIZE}px solid;
    border-right-color: inherit;
    left: -${ARROW_SIZE - 1}px;
  `,
    left: `
      ${sideBorders}
      border-left: ${ARROW_SIZE}px solid;
      border-left-color: inherit;
      right: -${ARROW_SIZE - 1}px;
    `,
    bottom: `
      ${upDownBorders}
      border-bottom: ${ARROW_SIZE}px solid;
      border-bottom-color: inherit;
      top: -${ARROW_SIZE - 1}px;
    `,
  },
  tipPosition: {
    top: `
      left: ${PERCENTAGE_X}%;
      bottom: ${tipPositionDistance};
      transform: translateX(-${PERCENTAGE_X}%);
    `,
    right: `
      left: ${tipPositionDistance};
      top: ${PERCENTAGE_X}%;
      transform: translateY(-${PERCENTAGE_X}%);
  `,
    left: `
      right: ${tipPositionDistance};
      top: ${PERCENTAGE_X}%;
      transform: translateY(-${PERCENTAGE_X}%);
    `,
    bottom: `
      left: ${PERCENTAGE_X}%;
      top: ${tipPositionDistance};
      transform: translateX(-${PERCENTAGE_X}%);
    `,
  },
};

export default placementConfig;
