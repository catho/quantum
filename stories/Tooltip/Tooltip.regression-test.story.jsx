import React from 'react';
import { storiesOf } from '@storybook/react';

import { Tooltip, Button, Row, Col } from '../../components';

const directions = ['top', 'bottom', 'left', 'right'];

const tooltipExampleButton = direction => (
  <Row style={{ padding: '50px 180px' }}>
    <Col>
      <Tooltip placement={direction} text="This is a hint." visible>
        <Button>Hover me</Button>
      </Tooltip>
    </Col>
  </Row>
);

directions.forEach(direction => {
  storiesOf('Tooltip', module).add(`${direction}`, () =>
    tooltipExampleButton(direction),
  );
});
