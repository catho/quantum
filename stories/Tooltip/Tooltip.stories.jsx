import { Row, Col, Tooltip } from '../../components';

export default {
  title: 'Tooltip',
  component: Tooltip,
};

const Template = (args) => (
  <Row style={{ paddingTop: '72px' }}>
    <Col>
      <Tooltip text="Text..." {...args}>
        Tooltip trigger
      </Tooltip>
    </Col>
  </Row>
);

export const Default = Template.bind({});

export const Multiline = Template.bind({});
Multiline.args = {
  text: 'Multiline\ntext...',
  multiline: true,
};

export const WithPlacement = Template.bind({});
WithPlacement.args = {
  placement: 'right',
};

export const Visible = Template.bind({});
Visible.args = {
  visible: true,
};
