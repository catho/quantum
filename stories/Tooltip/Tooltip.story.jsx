import React from 'react';
import { storiesOf } from '@storybook/react';
import {
  AutoExample,
  Tab,
  Title,
  StoryContainer,
  SimpleHighlight,
} from '@catho/quantum-storybook-ui';
import { Row, Col } from '../../components/Grid';
import Tooltip from '../../components/Tooltip';
import examples from './examples/examples';

const TabExample = (
  <Tab title="Examples">
    <StoryContainer>
      <Title as="h2">Examples</Title>
      {examples.map(example => (
        <Row key={example.code}>
          <Col xsmall={4} small={8} medium={6}>
            <SimpleHighlight>{example.code}</SimpleHighlight>
          </Col>
          <Col xsmall={4} small={8} medium={6}>
            <Row>
              <Col xsmall={10} small={8} medium={6}>
                {example.component}
              </Col>
            </Row>
          </Col>
        </Row>
      ))}
    </StoryContainer>
  </Tab>
);

const description = `Tooltips provide additional information upon hover or focus.
They often contain helper text that is contextual to an element.`;

storiesOf('Tooltip', module).add('Basic', () => (
  <AutoExample
    description={description}
    component={Tooltip}
    componentProps={{
      children: 'Hover me',
      text: 'This is a hint.',
    }}
    additionalTabs={TabExample}
  />
));
