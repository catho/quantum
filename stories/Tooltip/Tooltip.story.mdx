import { Meta, Canvas, Story, ArgsTable } from '@storybook/addon-docs';

import Tooltip from '../../components/Tooltip';
import {
  Default,
  Multiline,
  WithPlacement,
  Visible,
} from './Tooltip.stories.jsx';
import { Header } from '../shared';

<Meta
  title="Tooltip"
  component={Tooltip}
/>

<Header title="Tooltip">
  Tooltips provide additional information upon hover or focus. They often
  contain helper texts that are related to an element.
</Header>

- [Importing and usage](#importing)
- [API](#api)
- [Examples](#examples)

## <a id="importing"></a>Importing and usage

```js
import Tooltip from '@catho/quantum/Tooltip';
```

<Canvas>
  <Story story={Default} />
</Canvas>

## <a id="api"></a>API

<ArgsTable of={Tooltip} />

## <a id="examples"></a>Examples

### Multiline

<Canvas>
  <Story story={Multiline} />
</Canvas>

### With placement

<Canvas>
  <Story story={WithPlacement} />
</Canvas>

### Visible

<Canvas>
  <Story story={Visible} />
</Canvas>
