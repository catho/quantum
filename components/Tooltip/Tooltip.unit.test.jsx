import React from 'react';
import { mount } from 'enzyme';
import Tooltip from './Tooltip';

const TOOLTIP_TEXT = 'This is a hint';
const MULTILINE_TOOLTIP_TEXT = `Lorem ipsum dolor sit amet, 
consectetur adipiscing elit. 
Aenean bibendum facilisis sem viverra fringilla. 
Suspendisse finibus libero nec justo semper.`;

describe('Tooltip component ', () => {
  describe('All positions', () => {
    it('Should match the snapshot when place is top', () => {
      const tooltip = mount(
        <Tooltip place="top" text={TOOLTIP_TEXT}>
          Hover Me
        </Tooltip>,
      );
      expect(tooltip.html()).toMatchSnapshot();
    });

    it('Should match the snapshot when place is right', () => {
      const tooltip = mount(
        <Tooltip place="right" text={TOOLTIP_TEXT}>
          Hover Me
        </Tooltip>,
      );
      expect(tooltip.html()).toMatchSnapshot();
    });

    it('Should match the snapshot when place is bottom', () => {
      const tooltip = mount(
        <Tooltip place="bottom" text={TOOLTIP_TEXT}>
          Hover Me
        </Tooltip>,
      );
      expect(tooltip.html()).toMatchSnapshot();
    });

    it('Should match the snapshot when place is left', () => {
      const tooltip = mount(
        <Tooltip place="left" text={TOOLTIP_TEXT}>
          Hover Me
        </Tooltip>,
      );
      expect(tooltip.html()).toMatchSnapshot();
    });
  });
  describe('Multiline', () => {
    it('Should match the snapshot when is multiline', () => {
      const tooltip = mount(
        <Tooltip text={MULTILINE_TOOLTIP_TEXT} multiline>
          Hover Me
        </Tooltip>,
      );
      expect(tooltip.html()).toMatchSnapshot();
    });
    it('Should match the snapshot when is not multiline', () => {
      const tooltip = mount(
        <Tooltip text={MULTILINE_TOOLTIP_TEXT}>Hover Me</Tooltip>,
      );
      expect(tooltip.html()).toMatchSnapshot();
    });
  });
});
