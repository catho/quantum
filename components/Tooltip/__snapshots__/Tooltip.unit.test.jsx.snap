// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component  All positions Should match the snapshot when place is bottom 1`] = `"<div class=\\"Tooltip__Wrapper-sc-1s2fcfh-2 lmPKWo\\"><div class=\\"Tooltip__Tip-sc-1s2fcfh-0 dvvhdX\\"><span class=\\"Tooltip__TipText-sc-1s2fcfh-1 hhWQQw\\">This is a hint</span></div>Hover Me</div>"`;

exports[`Tooltip component  All positions Should match the snapshot when place is left 1`] = `"<div class=\\"Tooltip__Wrapper-sc-1s2fcfh-2 lmPKWo\\"><div class=\\"Tooltip__Tip-sc-1s2fcfh-0 dvvhdX\\"><span class=\\"Tooltip__TipText-sc-1s2fcfh-1 hhWQQw\\">This is a hint</span></div>Hover Me</div>"`;

exports[`Tooltip component  All positions Should match the snapshot when place is right 1`] = `"<div class=\\"Tooltip__Wrapper-sc-1s2fcfh-2 lmPKWo\\"><div class=\\"Tooltip__Tip-sc-1s2fcfh-0 dvvhdX\\"><span class=\\"Tooltip__TipText-sc-1s2fcfh-1 hhWQQw\\">This is a hint</span></div>Hover Me</div>"`;

exports[`Tooltip component  All positions Should match the snapshot when place is top 1`] = `"<div class=\\"Tooltip__Wrapper-sc-1s2fcfh-2 lmPKWo\\"><div class=\\"Tooltip__Tip-sc-1s2fcfh-0 dvvhdX\\"><span class=\\"Tooltip__TipText-sc-1s2fcfh-1 hhWQQw\\">This is a hint</span></div>Hover Me</div>"`;
